### ISSUE TITLE
[Question] about penalty term

### ISSUE THREAD

**Issue by mayer79** [microsoft/LightGBM#2177]
I was going through the code of "feature_histogram.hpp" and got a question about the penalty terms in the loss function. Its gain is evaluated in
```
static double GetLeafSplitGainGivenOutput(double sum_gradients, double sum_hessians, double l1, double l2, double output) {
    const double sg_l1 = ThresholdL1(sum_gradients, l1);
    return -(2.0 * sg_l1 * output + (sum_hessians + l2) * output * output);
  }
```
The `output` is the leaf value that incorporates both the penalizing effect of reg_alpha (l1) and reg_lambda (l2). 

My question: Where is the l1 penalty being added to the gain? I'd expect a `+ l1 * abs(output)` somewhere. Maybe I am not looking at the right code piece?

**Comment by guolinke** [comment]
@mayer79 it is not linear model or neural networks, the l1 is not added into gain directly.
The L1 and L2 are applied into leaf outputs, then we use "regularized" outputs to calculate split gains.

**Comment by mayer79** [comment]
@guolinke hmm, bhe l2 penalty is added to the gain (`output * output * l2), so I was expecting the same for l1. XGBoost adds it, see the answer to my question https://github.com/dmlc/xgboost/issues/4464

**Comment by guolinke** [comment]
@mayer79 this function is introduced in this PR:
it is another form of:
```
double abs_sum_gradients = std::fabs(sum_gradients);
double reg_abs_sum_gradients = std::max(0.0, abs_sum_gradients - l1);
return (reg_abs_sum_gradients * reg_abs_sum_gradients) / (sum_hessians + l2);
```
which is the same as:
https://github.com/dmlc/xgboost/blob/17df5fd296aaff18123c53a777f24111b3dd7336/src/tree/split_evaluator.cc#L145


for another form of xgboost:
```
weight * (2.0 * stats.sum_grad + stats.sum_hess * weight
        + params_.reg_lambda * weight)
        + 2.0 * params_.reg_alpha * std::abs(weight);
```
where weight is 
```
-ThresholdL1(stats.sum_grad, params_.reg_alpha) / (stats.sum_hess + params_.reg_lambda);
```

Combine them, you will find it is the same as the one used in lightgbm.

**Comment by mayer79** [comment]
The output resp. leaf weights are definitively similar, fully agree with that. But are the split gains? The additive term `params_.reg_alpha * std::abs(weight)` is difficult to track in lgb. I will need paper and pen ;-).

**Comment by mayer79** [comment]
@guolinke: going again through the calculations showed that you are completely (and obviously) right. Thx for the explanations.
